### DISCUSSION TITLE
Is it possible to add a TextBox to eclipse GEF Palette?

### DISCUSSION THREAD

**Discussion by pramakan-tibco** [eclipse-gef/gef-classic#448] [Q&A]
Hi,
I'm developing a Graphical Editor in eclipse . I created a palette with palette entries , by default palette is displayed a scroll-Up and scroll-down arrows to scroll.Now I want to add a TextBox to palette for searching the list of palette entries on palette.I attached an image in that I'm showing my requirement, where I want to place a textbox.
![Edited_2](https://github.com/eclipse/gef-classic/assets/107627949/e645e439-7469-4a56-bafa-da2d639a120e)

**Comment by azoitl** [comment]
I think this should be possible, but it will require a bit of tinkering. The palette itself is also a kind of GEF Viewer. So every entry in the palette has an EditPart. If I saw it correctly the EditPart for such a group is the `DrawerEditPart`. You could override this provide a figure where you have a search field and everytime the user enters a text call `refreshChildren()`. In `getModelChildren()` you can then filter the children to be shown. 

I hope this helps.

**Reply by ptziegler** [reply]
Keep in mind that GEF doesn't really have a concept for a text field (or any editable field), so the new figure would have to wrap an SWT widget.

The rest should then be straightforward: Hook a listener to the widget and then call refresh, just as you mentioned.

**Reply by pramakan-tibco** [reply]
@azoitl , @ptziegler Thanks for the reply,
@ptziegler I wanted to add this text field in palette, Any reference where I can create a editable figure in palette.

**Reply by azoitl** [reply]
One example, although not perfectly what you need, is the `LogicLabelEditManager` in  the LogicEditor example. There a text field is shown for direct editing of labels in the editor. I guess a similar approach could work here.

**Reply by ptziegler** [reply]
>  I wanted to add this text field in palette, Any reference where I can create a editable figure in palette.

Not a reference but just the general workflow:

    The figure needs to know the canvas it is painted in, in order to create the text field.
    The text field should be created whenever the figure is painted.
    The text field should be disposed whenever the figure is disposed.
    The text field should be resized whenever the figure is resized.

And of course you have to deal with fun stuff like scrolling.
